docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+```
+
+- Generate CA certificates if this is the first control plane machine in the cluster
+- Generate the kubeconfig secret if this is the first control plane machine in the cluster
+
+Given the following `cluster.yaml` file:
+
+```yaml
+kind: Cluster
+apiVersion: cluster.x-k8s.io/v1alpha3
+metadata:
+  name: my-cluster
+  namespace: default
+Spec:
+    clusterNetwork:
+    pods:
+      cidrBlocks: ["192.168.0.0/16"]
+  controlPlaneRef:
+    kind: ControlPlane
+    apiVersion: cluster.x-k8s.io/v1alpha3
+    name: my-controlplane
+    namespace: default
+  infrastructureRef:
+    kind: AWSCluster
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+    name: my-awscluster
+    namespace: default
+---
+kind: ControlPlane
+apiVersion: cluster.x-k8s.io/v1alpha3
+metadata:
+  name: my-control-plane
+  namespace: default
+spec:
+  replicas: 1
+  version: v1.16.0
+  infrastructureTemplate:
+    kind: AWSMachineTemplate
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+    namespace: default
+    name: my-awsmachinetemplate
+  kubeadmConfigSpec:
+    initConfiguration:
+      nodeRegistration:
+        name: '{{ ds.meta_data.hostname }}'
+        kubeletExtraArgs:
+          cloud-provider: aws
+    clusterConfiguration:
+      apiServer:
+        extraArgs:
+          cloud-provider: aws
+      controllerManager:
+        extraArgs:
+          cloud-provider: aws
+    joinConfiguration:
+      controlPlane: {}
+      nodeRegistration:
+        name: '{{ ds.meta_data.hostname }}'
+        kubeletExtraArgs:
+          cloud-provider: aws
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: AWSMachineTemplate
+metadata:
+  name: my-awsmachinetemplate
+  namespace: default
+spec:
+  ami:
+    id: ami-123456abcdef
+  instanceType: t3.large
+  iamInstanceProfile: "control-plane.cluster-api-provider-aws.sigs.k8s.io"
+  sshKeyName: my-ssh-key
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: AWSCluster
+metadata:
+  name: my-awscluster
+  namespace: default
+spec:
+  region: us-east-1
+```
+
+![controlplane-init-1](images/controlplane/controlplane-init-1.png)
+![controlplane-init-2](images/controlplane/controlplane-init-2.png)
+![controlplane-init-3](images/controlplane/controlplane-init-3.png)
+![controlplane-init-4](images/controlplane/controlplane-init-4.png)
+
+##### Scale Up
+
+- Allow scale up a control plane with stacked etcd to only 1, 3, 5, or 7, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
+- However, allow a control plane using an external etcd cluster to scale up to other numbers such as 2 or 4.
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
```

Comment:
```suggestion
- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
```

This stanza exists elsewhere in the doc too - do we need it in both places?

Reply:
Not likely.